src/hotspot/share/cds/cdsConfig.cpp

```diff
@@ -120,6 +120,26 @@ void CDSConfig::ergo_initialize() {
   if (!is_dumping_heap()) {
     _is_dumping_full_module_graph = false;
   }
+
+#ifdef _LP64
+  //
+  // By default, when using AOTClassLinking, use the CompressedOops::HeapBasedNarrowOop
+  // mode so that AOT code can be always work regardless of runtime heap range.
+  //
+  // If you are *absolutely sure* that the CompressedOops::mode() will be the same
+  // between training and production runs (e.g., if you specify -Xmx128m for
+  // both training and production runs, and you know the OS will always reserve
+  // the heap under 4GB), you can explicitly disable this with:
+  //     java -XX:+UnlockDiagnosticVMOptions -XX:-UseCompatibleCompressedOops ...
+  // However, this is risky and there's a chance that the production run will be slower
+  // than expected because it is unable to load the AOT code cache.
+  //
+  if (UseCompressedOops && AOTCodeCache::is_caching_enabled()) {
+    FLAG_SET_ERGO_IF_DEFAULT(UseCompatibleCompressedOops, true);
+  } else if (!FLAG_IS_DEFAULT(UseCompatibleCompressedOops)) {
+    FLAG_SET_ERGO(UseCompatibleCompressedOops, false);
```

Comment:
Why do we need to set it to false explicitly here? If it is not default shouldn't we use whatever value the user specified?

Reply:
I consider the flag make sense only for AOT code and not for general usage. See description: [globals.hpp#L132](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/runtime/globals.hpp#L132)

We don't want generate more complex encoding if we don't generate/use AOT code. Or compressed oops are off.
We had also recently discussed how we should disable some AOT features if AOT caching is off - the conclusion was that we should silently ignore them (may be produce UL output on demand). That is why I used ERGO setting.

Reply:
okay